File: sudoku.c
//  File: sudoku.c              //
//  Date: 15th of Dec. 2021     //
//  Version: 1.0                //
//  Goal: practice backtracking //

#include <stdio.h>
#include <stdbool.h>


//grid: input given sudoku to solve
short int grid[9][9] = {
    {0,0,0,0,0,8,4,0,0},
    {0,3,0,0,6,0,0,0,9},
    {0,0,0,0,2,0,7,0,0},
    {6,0,0,4,0,0,0,3,0},
    {0,9,0,0,1,0,0,2,0},
    {0,8,0,0,0,9,0,0,6},
    {0,0,0,0,0,0,0,0,0},
    {8,0,0,0,3,0,0,5,0},
    {0,0,2,0,0,6,0,0,0}
};

bool solve();
bool isValid(int row, int col, int value);
bool sameRow(int row, int value);
bool sameCol(int col, int value);
bool sameSquare(int row, int col, int value);
void show();

int main() {
    fprintf(stdout, "\n\033[43m\033[30mHere's your input:\033[0m\n\n"); //using colors for bash
    show();
    bool solved = solve();
    if(solved) {
        fprintf(stdout, "\n\033[42m\033[30mSolution found:\033[0m\n\n");
        show();
    } else {
        fprintf(stderr, "\n\033[41m\033[30mError. No solution found.\033[0m\n\n");
    }
}

//recursive function to solve the sudoku
bool solve() {
    bool all = true;
    for(int row=0;row<9;row++) {
        for(int col=0;col<9;col++) {
            if(grid[row][col]==0) all = false;
        }
    }
    if(all == true) return true;

    for(int row=0;row<9;row++) {
        for(int col=0;col<9;col++) {
            if(grid[row][col]==0) { //only for non-given elements
                for(grid[row][col]=1;grid[row][col]<=9;grid[row][col]++) {
                    if(isValid(row,col,grid[row][col]) && solve()) return true; //everything correct --> recursion
                }
                grid[row][col] = 0;
                return false;
            }
        }
    } 
    return false;
}

//checks if the number would be valid
bool isValid(int row, int col, int value) {
    if(sameRow(row, value) && sameCol(col, value) && sameSquare(row, col, value)) return true;
    return false;
}

//checks row
bool sameRow(int row, int value) {
    short int counter = 0;
    for(int col=0;col<9;col++) {
        if(grid[row][col] == value) counter++;
    }
    if(counter>1) {
        return false;
    }
    return true;
}

//checks column
bool sameCol(int col, int value) {
    short int counter = 0;
    for(int row=0;row<9;row++) {
        if(grid[row][col] == value) counter++;
    }
    if(counter>1) {
        return false;
    }
    return true;
}

//checks 3x3-squares
bool sameSquare(int row, int col, int value) {
    short int counter = 0;
    for(int r=(row/3)*3;r<(row/3)*3+3;r++) {
        for(int c=(col/3)*3;c<(col/3)*3+3;c++) {
            if(grid[r][c]==value) counter++;
        }
    }
    if(counter>1) {
        return false;
    }
    return true;
}

//prints the grid
void show() {
    for(int row=0;row<9;row++) {
        for(int col=0;col<9;col++) {
            fprintf(stdout, "%d ", grid[row][col]);
            (((col % 3)-2)==0 && col!=8) ? (fprintf(stdout, "| ")):(fprintf(stdout, ""));
        }
        if((row%3)-2==0 && row!=8) {
            fprintf(stdout, "\n");
            for(int count=0;count<21;count++) {
                fprintf(stdout, "-");
            }
        }
        fprintf(stdout, "\n");
    }
    fprintf(stdout, "\n");
}

File: sudoku.py
# sudoku.py
# 1st of January 2022


def main() -> None:
    grid: list = [[0, 0, 0, 8, 0, 5, 0, 1, 3],
                  [9, 0, 5, 2, 0, 3, 6, 0, 0],
                  [6, 0, 3, 0, 9, 0, 2, 0, 4],
                  [0, 0, 1, 0, 0, 0, 0, 0, 5],
                  [0, 4, 0, 1, 0, 0, 7, 0, 6],
                  [2, 5, 6, 3, 0, 4, 8, 9, 0],
                  [5, 9, 0, 0, 0, 7, 1, 0, 2],
                  [1, 0, 2, 0, 8, 0, 4, 7, 0],
                  [0, 0, 4, 9, 1, 0, 0, 3, 8]]
    if solve(grid):
        for row in grid:
            print(row)
    else:
        print("Sorry. No solution found.")


def solve(grid: list) -> bool:
    count_zeros: int = 0
    for row in range(9):
        for col in range(9):
            if grid[row][col] == 0:
                count_zeros += 1
    if count_zeros == 0:
        return True
    for row in range(9):
        for col in range(9):
            if grid[row][col] == 0:
                for value in range(1, 10):
                    grid[row][col] = value
                    if is_valid(grid, value, row, col) and solve(grid):
                        return True
                    else:
                        grid[row][col] = 0
                return False
    return False


def is_valid(grid: list, value: int, row: int, col: int) -> bool:
    if same_column(grid, value, col) or same_row(grid, value, row) or same_square(grid, value, row, col):
        return False
    return True


def same_column(grid: list, value: int, col: int) -> bool:
    counter: int = 0
    for row in range(9):
        if grid[row][col] == value:
            counter += 1
    if counter < 2:
        return False
    return True


def same_row(grid: list, value: int, row: int) -> bool:
    counter: int = 0
    for col in range(9):
        if grid[row][col] == value:
            counter += 1
    if counter < 2:
        return False
    return True


def same_square(grid: list, value: int, row: int, col: int) -> bool:
    row_start: int = (row // 3) * 3
    col_start: int = (col // 3) * 3
    counter: int = 0
    for row_counter in range(row_start, row_start + 3):
        for col_counter in range(col_start, col_start + 3):
            if grid[row_counter][col_counter] == value:
                counter += 1
    if counter < 2:
        return False
    return True


main()
